### PR TITLE
Account for moving channel state to flush complete in timeout processing

### PR DESCRIPTION
**Pull request by chatton**
## Description

closes: #4262

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [x] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/core/04-channel/keeper/timeout_test.go
Comment (damiannolan) [inline modules/core/04-channel/keeper/timeout_test.go:1]: Awesome, these tests already are setup with `expResult`! 

Would also be nice to have a test where the channel remains in `FLUSHING` after a timeout but upgrade timeout is valid. Essentially having an additional in-flight packet so that the upgrade timeout logic is checked but the channel state transition doesn't occur because there is another commitment in state

File: modules/core/04-channel/keeper/timeout.go
Comment (damiannolan) [inline modules/core/04-channel/keeper/timeout.go:137]: This godoc is a little funky sounding atm! 

> Otherwise, if the channel is in if there are no more inflight packets

File: modules/core/04-channel/keeper/timeout.go
Comment (damiannolan) [inline modules/core/04-channel/keeper/timeout.go:191]: acknowledged, timed out 

probs better to just say flushed for this one 

```suggestion
			// set the channel state to flush complete if all packets have been flushed.
```

File: modules/core/04-channel/keeper/timeout.go
Comment (damiannolan) [inline modules/core/04-channel/keeper/timeout.go:186]:
```suggestion
				// committing an error receipt to state, restoring the channel and successfully timing out the packet.
```

File: modules/core/04-channel/keeper/timeout.go
Comment (damiannolan) [inline modules/core/04-channel/keeper/timeout.go:162]: This could be slightly funky.

Does `channe.Close()` mutate the channel (I assume so). 

Should this check be after the flushing state checks.
Otherwise you'll close the channel with a timed out packet on an ordered channel, and have this upgrade info laying around in store.

**Review by damiannolan — COMMENTED** [review]
Just nits on comments and such. 
Would like to add that one extra test case before merging this! 

Super clean work as always ❤️

File: modules/core/04-channel/keeper/timeout.go
Comment (chatton) [reply modules/core/04-channel/keeper/timeout.go:162]: yeah I think you're right! We cane move this to after the new flushing check, nice catch

**Review by damiannolan — APPROVED** [review]
LGTM! Pending what we discussed sync.

- Link issue we created
- Linting

🚀 🚀 🚀

**Pull request merged**
